### ISSUE TITLE
TeXLive package manager requires an upgrade!

### ISSUE THREAD

**Issue by flowgunso** [sharelatex/sharelatex-docker-image#90]
Trying to use the TexLive package manager on `sharelatex/sharelatex` or `sharelatex/sharelatex:latest` with:
`tlmgr update --self --all` or `tlmgr install scheme-full` both results in the following message:
```
tlmgr: Remote repository is newer than local (2017 < 2018)
Cross release updates are only supported with
  update-tlmgr-latest(.sh/.exe) --update
Please see https://tug.org/texlive/upgrade.html for details.
```

**Comment by magic-lantern** [comment]
Also seeing the same error. Not sure what the recommended next steps - please advise!

**Comment by magic-lantern** [comment]
Just wanted to clarify - because tlmgr is outdated, it is not possible to install any packages.

**Comment by cemulate** [comment]
I'm experiencing this issue too. I tried to rebuild the docker image with @magic-lantern's PR #91, but ran into problems in the build.

In the mean time, I was able to `docker exec -it sharelatex bash` and follow [the TeX Live update instructions](https://www.tug.org/texlive/upgrade.html) as a workaround for now.

----

**Edit**: This requires significant finagling with `sv` to get the `sharelatex` processes to use the new `PATH` in the environment.

**Comment by magic-lantern** [comment]
@cemulate - Can you provide details about what error you get in the build process? Although I'm not on the core team for Sharelatex, I'd like to see if I can resolve that error and/or improve PR #91

**Comment by cemulate** [comment]
@magic-lantern It's almost certainly me not knowing how to build `sharelatex`, or an existing problem. I cloned your master branch, went in, and typed `make`, successfully building `sharelatex-base` (that makes me think it has nothing to do with Tex Live).

Then I typed `make build-community` and it fails deep into step 16/22 with (don't mind the color code characters)

```
[0m[91mCloning into 'launchpad'...
[0mLoading "Gruntfile.coffee" tasks...[31mERROR[39m
[31m>> [39mError: Cannot find module 'es6-promise'
[33mWarning: Task "compile" not found. Use --force to continue.[39m
 
[31mAborted due to warnings.[39m
The command '/bin/sh -c cd /var/www/sharelatex;     npm install;    grunt install;  bash -c 'source ./bin/install-services';    cd web;     npm install;    npm install bcrypt;     cd modules;     git clone https://github.com/sharelatex/launchpad-web-module.git launchpad;     grunt compile;' returned a non-zero code: 3
Makefile:9: recipe for target 'build-community' failed
make: *** [build-community] Error 3
```

[Pastebin output here](https://pastebin.com/dg3En0tc) if anyone wants to take a look. 

Perhaps I don't know how to build `sharelatex` properly?

**Comment by magic-lantern** [comment]
@cemulate  So what I did on macOS Sierra (10.12.6) and Ubuntu 16.04.4 LTS to build the docker containers was to first delete any existing sharelatex docker images and then run these two commands from the Makefile:

```
docker build -f Dockerfile-base -t sharelatex/sharelatex-base .
docker build -f Dockerfile -t sharelatex/sharelatex .
```

When I ran those two steps (sequentially - first one has to finish before you start the second) there were quite a few warnings due to various outdated node/NPM dependencies, but everything worked in the end.

Your error appears to be due to problems with the NPM build steps; not related to TexLive as you said.

Perhaps the /var/www/sharelatex/web/npm-debug.log log file referenced may have more details on the cause of the error. Through a quick google search, problems similar to this have come up with other projects - not sure though if it is due to outdated Node/NPM dependencies, insufficient disk space, networking error, or the NPM CDN having a problem.

**Comment by beenje** [comment]
@cemulate, I think the issue is that the latest version of the sharelatex git repo is used in the docker image.
The build of the docker image should be reproducible. I just opened a PR to pin the version used: https://github.com/sharelatex/sharelatex-docker-image/pull/94

**Comment by cemulate** [comment]
@beenje that makes sense. 

If anyone comes around to this issue, I've been using [xuio/sharelatex-docker-image-full](https://hub.docker.com/r/xuio/sharelatex-docker-image-full/) in the meantime, until this image can be updated to use the latest TeX Live.

**Comment by crockeea** [comment]
I was able to install TexLive2017 using:

>docker exec sharelatex tlmgr option repository ftp://tug.org/historic/systems/texlive/2017/tlnet-final
>docker exec sharelatex tlmgr install scheme-full

[Note: this no longer works, because tug stopped hosting the historic TeXLive packages.]

**Comment by georgekamel** [comment]
Another option is to add the following PATH variable to the sharelatex part of the docker-compose.yml file before starting the containers:

```
PATH: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/2018/bin/x86_64-linux/
```

Then you can quite easily follow the [TeX Live update instructions](https://www.tug.org/texlive/upgrade.html), disregarding steps 5 and 6.

**Comment by phillijw** [comment]
@georgekamel One issue I had with those instructions was step 11:

```
root@e04b4e037152:/usr/local/texlive/2018# luaotfload-tool -fu
bash: luaotfload-tool: command not found
```

I just ended up skipping it. No idea if that will cause issues or not

**Comment by MadMud** [comment]
Hi, I have pasted my workaround to this to the main repo https://github.com/sharelatex/sharelatex/issues/601. Perhaps I should move it here? Anyway the to;Dr is: do not update reinstall texlife in a daughter image.

**Comment by tngTUDOR** [comment]
> @georgekamel One issue I had with those instructions was step 11:
> 
> ```
> root@e04b4e037152:/usr/local/texlive/2018# luaotfload-tool -fu
> bash: luaotfload-tool: command not found
> ```
> 
> I just ended up skipping it. No idea if that will cause issues or not

The `luaotfload-tool` tool is part of the luaotfload package [CTAN repo for luaotfload](https://ctan.org/tex-archive/macros/luatex/generic/luaotfload?lang=en).
You can install it with:
```bash
docker exec sharelatex tlmgr install luaotfload
```

**Comment by markedphillips** [comment]
https://gist.github.com/markedphillips/3d566d3721f8c80e313d6ce365cd87a0

**Comment by markedphillips** [comment]
Wow, what a load of annoyingness. Here is an updated [docker-compose.yml ](https://gist.github.com/markedphillips/3d566d3721f8c80e313d6ce365cd87a0)file. Put it in a folder called overleaf. Then "docker-compose up  -d" and it will pull an updated latex installation.

**Comment by GuijonGustavo** [comment]
My recomendation is:

1. Enter to container: $sudo docker container exec -it sharelatex /bin/bash
2. Inside the container: # tlmgr install scheme-full
3. Then: # exit
4. Finally: $sudo docker commit -m "installing all latex packages" [CONTAINER ID] sharelatex/sharelatex:v1

**Comment by mserranom** [comment]
We’ve merged this repository into https://github.com/overleaf/overleaf, so we’re closing all of its issues and will archive it. If this is still a problem, please open a new issue on https://github.com/overleaf/overleaf . Thanks!
